File: android-app/app/src/main/kotlin/com/mysplitwise/app/network/SyncApi.kt
package com.mysplitwise.app.network

import com.mysplitwise.app.model.Expense
import com.mysplitwise.app.model.ExpenseShare
import com.mysplitwise.app.model.Group
import com.mysplitwise.app.model.GroupType
import com.mysplitwise.app.model.LineItem
import com.mysplitwise.app.model.NotificationPrefs
import com.mysplitwise.app.model.PaymentMethod
import com.mysplitwise.app.model.PullResponse
import com.mysplitwise.app.model.User
import io.ktor.http.HttpMethod
import kotlinx.serialization.Serializable

/** Typed wrappers around every backend call Phase 1 needs — port of `src/lib/sync-api.ts`. */
object SyncApi {
    // ---- Sync ----

    suspend fun pullState(): PullResponse = ApiClient.request(HttpMethod.Get, Endpoints.SYNC_PULL)

    @Serializable private data class OkResponse(val ok: Boolean)
    @Serializable private data class EmptyBody(val placeholder: Boolean? = null)

    suspend fun claimInvites() {
        ApiClient.request<EmptyBody, OkResponse>(HttpMethod.Post, Endpoints.SYNC_CLAIM_INVITES, EmptyBody())
    }

    @Serializable private data class LogActivityBody(val eventType: String)
    suspend fun logActivity(eventType: String) {
        ApiClient.request<LogActivityBody, OkResponse>(HttpMethod.Post, Endpoints.LOG_ACTIVITY, LogActivityBody(eventType))
    }

    // ---- Friends ----

    @Serializable private data class AddFriendBody(val name: String, val email: String)
    @Serializable data class AddFriendResponse(val status: String, val friend: User)

    suspend fun addFriend(name: String, email: String): AddFriendResponse =
        ApiClient.request(HttpMethod.Post, Endpoints.FRIENDS, AddFriendBody(name, email))

    suspend fun removeFriend(id: String) {
        ApiClient.request<OkResponse>(HttpMethod.Delete, Endpoints.friend(id))
    }

    // ---- Profile ----

    /** All-optional; only non-null fields are serialized (see `encodeDefaults = false` note below). */
    @Serializable
    data class ProfilePatch(
        val name: String? = null,
        val avatarColor: String? = null,
        val venmo: String? = null,
        val paypal: String? = null,
        val cashapp: String? = null,
        val baseCurrency: String? = null,
        val notificationsReadAt: String? = null,
        val onboarded: Boolean? = null,
        val phone: String? = null,
        val notificationPrefs: NotificationPrefs? = null,
    )

    @Serializable private data class UserResponse(val user: User)

    suspend fun updateProfile(id: String, patch: ProfilePatch): User =
        ApiClient.request<ProfilePatch, UserResponse>(HttpMethod.Patch, Endpoints.profile(id), patch).user

    // ---- Groups ----

    @Serializable private data class CreateGroupBody(val name: String, val type: GroupType, val memberIds: List<String>)
    @Serializable private data class GroupResponse(val group: Group)

    suspend fun createGroup(name: String, type: GroupType, memberIds: List<String>): Group =
        ApiClient.request<CreateGroupBody, GroupResponse>(HttpMethod.Post, Endpoints.GROUPS, CreateGroupBody(name, type, memberIds)).group

    @Serializable
    data class UpdateGroupBody(
        val name: String? = null,
        val type: GroupType? = null,
        val simplifyDebts: Boolean? = null,
        val monthlyBudget: Double? = null,
        val memberIds: List<String>? = null,
    )

    suspend fun updateGroup(id: String, patch: UpdateGroupBody): Group =
        ApiClient.request<UpdateGroupBody, GroupResponse>(HttpMethod.Patch, Endpoints.group(id), patch).group

    suspend fun deleteGroup(id: String) {
        ApiClient.request<OkResponse>(HttpMethod.Delete, Endpoints.group(id))
    }

    // ---- Expenses ----

    @Serializable
    data class NewExpense(
        val description: String,
        val amount: Double,
        val currency: String,
        val category: String,
        val date: String,
        val groupId: String? = null,
        val isSettlement: Boolean = false,
        val notes: String? = null,
        val shares: List<ExpenseShare>,
        val items: List<LineItem>? = null,
        val tax: Double? = null,
        val tip: Double? = null,
        val paymentMethod: PaymentMethod? = null,
    )

    @Serializable private data class ExpenseResponse(val expense: Expense)

    suspend fun createExpense(expense: NewExpense): Expense =
        ApiClient.request<NewExpense, ExpenseResponse>(HttpMethod.Post, Endpoints.EXPENSES, expense).expense

    @Serializable
    data class UpdateExpenseBody(
        val description: String? = null,
        val amount: Double? = null,
        val currency: String? = null,
        val category: String? = null,
        val date: String? = null,
        val groupId: String? = null,
        val notes: String? = null,
        val shares: List<ExpenseShare>? = null,
        val items: List<LineItem>? = null,
        val disputed: Boolean? = null,
        val disputeReason: String? = null,
    )

    suspend fun updateExpense(id: String, patch: UpdateExpenseBody): Expense =
        ApiClient.request<UpdateExpenseBody, ExpenseResponse>(HttpMethod.Patch, Endpoints.expense(id), patch).expense

    suspend fun deleteExpense(id: String) {
        ApiClient.request<OkResponse>(HttpMethod.Delete, Endpoints.expense(id))
    }

    // ---- Settlements ----

    @Serializable
    data class Payment(val fromId: String, val toId: String, val amount: Double, val currency: String, val groupId: String? = null)
    @Serializable private data class AddSettlementsBody(val payments: List<Payment>)
    @Serializable private data class AddSettlementsResponse(val expenses: List<Expense>)

    suspend fun addSettlements(payments: List<Payment>): List<Expense> =
        ApiClient.request<AddSettlementsBody, AddSettlementsResponse>(HttpMethod.Post, Endpoints.SETTLEMENTS, AddSettlementsBody(payments)).expenses

    // ---- Account ----

    suspend fun deleteAccount() {
        ApiClient.request<EmptyBody, OkResponse>(HttpMethod.Post, Endpoints.ACCOUNT_DELETE, EmptyBody())
    }
}

File: ios-app/MysplitwiseApp/Networking/SyncAPI.swift
import Foundation
import MysplitwiseCore

/// Typed wrappers around every `/api/*` call Phase 1 needs — port of
/// `src/lib/sync-api.ts`.
enum SyncAPI {
    // MARK: - Sync

    static func pullState() async throws -> PullResponse {
        try await APIClient.request("GET", Endpoints.syncPull)
    }

    struct OkResponse: Decodable { let ok: Bool }

    static func claimInvites() async throws {
        let _: OkResponse = try await APIClient.request("POST", Endpoints.syncClaimInvites, body: EmptyBody())
    }

    struct LogActivityBody: Encodable { let eventType: String }
    static func logActivity(eventType: String) async throws {
        let _: OkResponse = try await APIClient.request(
            "POST", Endpoints.logActivity, body: LogActivityBody(eventType: eventType)
        )
    }

    // MARK: - Friends

    struct AddFriendBody: Encodable { let name: String; let email: String }
    struct AddFriendResponse: Decodable { let status: String; let friend: User }

    static func addFriend(name: String, email: String) async throws -> AddFriendResponse {
        try await APIClient.request("POST", Endpoints.friends, body: AddFriendBody(name: name, email: email))
    }

    static func removeFriend(id: String) async throws {
        let _: OkResponse = try await APIClient.request("DELETE", Endpoints.friend(id))
    }

    // MARK: - Profile

    /// All-optional so `JSONEncoder` only serializes the fields you actually set
    /// (Swift's Codable synthesis uses `encodeIfPresent` for Optional properties).
    /// No `avatarUrl` field yet — avatar photo upload/clearing is Phase 2+.
    struct ProfilePatch: Encodable {
        var name: String?
        var avatarColor: String?
        var venmo: String?
        var paypal: String?
        var cashapp: String?
        var baseCurrency: String?
        var notificationsReadAt: String?
        var onboarded: Bool?
        var phone: String?
        var notificationPrefs: NotificationPrefs?
    }

    struct UserResponse: Decodable { let user: User }

    static func updateProfile(id: String, patch: ProfilePatch) async throws -> User {
        let response: UserResponse = try await APIClient.request("PATCH", Endpoints.profile(id), body: patch)
        return response.user
    }

    // MARK: - Groups

    struct CreateGroupBody: Encodable { let name: String; let type: GroupType; let memberIds: [String] }
    struct GroupResponse: Decodable { let group: Group }

    static func createGroup(name: String, type: GroupType, memberIds: [String]) async throws -> Group {
        let response: GroupResponse = try await APIClient.request(
            "POST", Endpoints.groups, body: CreateGroupBody(name: name, type: type, memberIds: memberIds)
        )
        return response.group
    }

    struct UpdateGroupBody: Encodable {
        var name: String?
        var type: GroupType?
        var simplifyDebts: Bool?
        var monthlyBudget: Double?
        var memberIds: [String]?
    }

    static func updateGroup(id: String, patch: UpdateGroupBody) async throws -> Group {
        let response: GroupResponse = try await APIClient.request("PATCH", Endpoints.group(id), body: patch)
        return response.group
    }

    static func deleteGroup(id: String) async throws {
        let _: OkResponse = try await APIClient.request("DELETE", Endpoints.group(id))
    }

    // MARK: - Expenses

    struct NewExpense: Encodable {
        var description: String
        var amount: Double
        var currency: String
        var category: String
        var date: String
        var groupId: String?
        var isSettlement: Bool = false
        var notes: String?
        var shares: [ExpenseShare]
        var items: [LineItem]?
        var tax: Double?
        var tip: Double?
        var paymentMethod: PaymentMethod?
    }

    struct ExpenseResponse: Decodable { let expense: Expense }

    static func createExpense(_ expense: NewExpense) async throws -> Expense {
        let response: ExpenseResponse = try await APIClient.request("POST", Endpoints.expenses, body: expense)
        return response.expense
    }

    struct UpdateExpenseBody: Encodable {
        var description: String?
        var amount: Double?
        var currency: String?
        var category: String?
        var date: String?
        var groupId: String?
        var notes: String?
        var shares: [ExpenseShare]?
        var items: [LineItem]?
        var disputed: Bool?
        var disputeReason: String?
    }

    static func updateExpense(id: String, patch: UpdateExpenseBody) async throws -> Expense {
        let response: ExpenseResponse = try await APIClient.request("PATCH", Endpoints.expense(id), body: patch)
        return response.expense
    }

    static func deleteExpense(id: String) async throws {
        let _: OkResponse = try await APIClient.request("DELETE", Endpoints.expense(id))
    }

    // MARK: - Settlements

    struct Payment: Encodable {
        let fromId: String
        let toId: String
        let amount: Double
        let currency: String
        let groupId: String?
    }
    struct AddSettlementsBody: Encodable { let payments: [Payment] }
    struct AddSettlementsResponse: Decodable { let expenses: [Expense] }

    static func addSettlements(_ payments: [Payment]) async throws -> [Expense] {
        let response: AddSettlementsResponse = try await APIClient.request(
            "POST", Endpoints.settlements, body: AddSettlementsBody(payments: payments)
        )
        return response.expenses
    }

    // MARK: - Account

    static func deleteAccount() async throws {
        let _: OkResponse = try await APIClient.request("POST", Endpoints.accountDelete, body: EmptyBody())
    }

    struct EmptyBody: Encodable {}
}
